driver/src/main/java/org/neo4j/driver/internal/InternalStatementResult.java

```diff
@@ -274,46 +248,34 @@ else if ( done )
     @Override
     public <T> List<T> list( Function<Record, T> mapFunction )
     {
-        if ( isEmpty() )
-        {
-            assertOpen();
-            return emptyList();
-        }
-        else if ( position == -1 && hasNext() )
+        if ( hasNext() )
         {
             List<T> result = new ArrayList<>();
+
             do
             {
                 result.add( mapFunction.apply( next() ) );
             }
             while ( hasNext() );
 
-            consume();
             return result;
         }
         else
         {
-            throw new ClientException(
-                    format( "Can't retain records when cursor is not pointing at the first record (currently at position %d)", position )
-            );
+            return emptyList();
         }
     }
 
-    @SuppressWarnings("StatementWithEmptyBody")
     @Override
     public ResultSummary consume()
     {
-        if(!open)
-        {
-            return summary;
-        }
-
         while ( !done )
         {
             connection.receiveOne();
+            recordBuffer.clear();
```

Comment:
Agree with @boggle. If we're draining records off the network, we shouldn't use the buffer so as to avoid unnecessary memory allocation.

The alternative (which might be a bit cleaner) would be to replace `receiveOne` with something like `discardOne` for this purpose. This would read the message off the network but not put it anywhere, thereby removing the need to purge the buffer each time. But unless this is quick and easy to do, we should probably stay with this for now and consider that alternative as an optimisation.